File: packages/core/src/utils/mcp.ts
import { Client } from '@modelcontextprotocol/sdk/client/index.js';
import { SSEClientTransport } from '@modelcontextprotocol/sdk/client/sse.js';
import { SecureLendError, AuthenticationError, NetworkError, ValidationError } from './errors';

interface MCPClientConfig {
  apiKey: string;
  mcpURL: string;
}

export class MCPClient {
  public mcp: Client;
  private config: MCPClientConfig;
  private debug: boolean = false;
  private isConnected: boolean = false;

  constructor(config: MCPClientConfig) {
    this.config = config;
    
    this.mcp = new Client(
      { name: '@securelend/mcp-client', version: '1.0.0' },
      { capabilities: { tools: {}, widgets: { supportsHtml: true } } }
    );
  }

  async connect(): Promise<void> {
    if (this.isConnected) return;

    try {
      if (this.debug) {
        console.log(`[SecureLend SDK] Connecting to MCP server at ${this.config.mcpURL}`);
      }
      const transport = new SSEClientTransport({
        url: this.config.mcpURL,
        headers: {
          'Authorization': `Bearer ${this.config.apiKey}`,
        },
      });
      await this.mcp.connect(transport);
      this.isConnected = true;
      if (this.debug) {
        console.log(`[SecureLend SDK] MCP client connected.`);
      }
    } catch (error: any) {
      this.isConnected = false;
      // Poor man's error mapping. The MCP SDK should have better error types.
      if (error.message?.includes('401')) {
        throw new AuthenticationError('MCP connection failed: Invalid API key.');
      }
      throw new NetworkError(`MCP connection failed: ${error.message}`);
    }
  }

  async callTool<T>(name: string, args: Record<string, any>): Promise<T> {
    await this.connect(); // Ensure connection exists

    if (this.debug) {
      console.log(`[SecureLend SDK] Calling tool: ${name} with args:`, args);
    }

    try {
      const result = await this.mcp.callTool({ name, arguments: args });
      
      if (this.debug) {
        console.log(`[SecureLend SDK] Tool result for ${name}:`, result);
      }

      // Assuming the relevant content is in the first part and is JSON.
      // This is a big assumption and might need to be more robust.
      const content = result.content[0];
      if (content?.type === 'resource' && content.resource.mimeType === 'application/json') {
        return JSON.parse(content.resource.text);
      } else if (content?.type === 'text') {
        // Attempt to parse if it looks like JSON
        try {
          return JSON.parse(content.text);
        } catch (e) {
            // Not JSON, return as is. This might not match <T>.
            return content.text as any;
        }
      }

      throw new SecureLendError('Unexpected tool response format from MCP server', 'mcp_error', result);
    } catch (error: any) {
      if (error instanceof SecureLendError) throw error;
      // Again, poor man's error mapping
      if (error.message?.includes('400')) {
        throw new ValidationError('Invalid tool arguments', error.details);
      }
      throw new SecureLendError(`Tool call failed: ${error.message}`, 'mcp_tool_error', error);
    }
  }

  setApiKey(apiKey: string): void {
    // MCP client doesn't support hot-swapping API key after connection.
    // This would require a disconnect/reconnect cycle.
    // For now, we'll just update the config for the *next* connection.
    this.config.apiKey = apiKey;
    this.isConnected = false; // Force reconnect on next call
  }

  enableDebug(): void {
    this.debug = true;
  }

  disableDebug(): void {
    this.debug = false;
  }
}

File: packages/python/securelend/utils/mcp.py
from typing import Any, Dict, Optional

from mcp.client import Client, ToolResult
from mcp.client.transports.http import StreamableHTTPClientTransport

from ..utils.errors import ServerError


class MCPClient:
    def __init__(self, *, api_key: str, mcp_url: str):
        self._api_key = api_key
        self._mcp_url = mcp_url
        self._debug = False
        self._is_connected = False
        self._mcp: Optional[Client] = None

    async def connect(self) -> None:
        if self._is_connected and self._mcp:
            return

        if self._debug:
            print(f"[SecureLend SDK] Connecting to MCP server at {self._mcp_url}")

        self._mcp = Client(
            name=f"securelend-python/{__import__('securelend').__version__}"
        )

        headers = {
            "User-Agent": f"securelend-python/{__import__('securelend').__version__}",
        }
        if self._api_key:
            headers["Authorization"] = f"Bearer {self._api_key}"

        transport = StreamableHTTPClientTransport(
            url=self._mcp_url,
            headers=headers,
        )
        self._mcp.add_transport(transport)

        try:
            await self._mcp.connect()
            self._is_connected = True
        except Exception as e:
            self._is_connected = False
            self._mcp = None
            raise ServerError(f"Failed to connect to MCP server: {e}") from e

    async def _ensure_connected(self) -> None:
        if not self._is_connected or not self._mcp:
            await self.connect()

    async def call_tool(self, name: str, args: Dict[str, Any]) -> Dict[str, Any]:
        await self._ensure_connected()
        assert self._mcp is not None

        if self._debug:
            print(f"[SecureLend SDK] Calling tool '{name}' with args: {args}")

        try:
            result: ToolResult = await self._mcp.tools.call(name, **args)
            return result.model_dump(by_alias=True)
        except Exception as e:
            raise ServerError(f"MCP tool call failed: {e}") from e

    def set_api_key(self, api_key: str) -> None:
        self._api_key = api_key
        self._is_connected = False
        self._mcp = None

    def enable_debug(self) -> None:
        self._debug = True

    def disable_debug(self) -> None:
        self._debug = False
